* Mining summary statistics with Racket

This module fetches summary statistics, starting with the new EBI API
described at https://www.ebi.ac.uk/gwas/summary-statistics and
https://www.ebi.ac.uk/gwas/rest/docs/api. The general idea is to start
from a gene (alias) or genome segment and find relevant SNPs and
associated phenotypes. These in turn we can use in GeneNetwork to
locate interesting hits related to mouse/rat genes and phenotypes.

This 'discovery' tool is based on Wikidata and EBI Ensembl API which
are both relatively stable resources.

In the near future we may be adding more resources. This work is part
of the GeneNetwork3 project.

Pjotr Prins

* EBI Summary Statistics

EBI is providing an update of their old GWAS resource in the form of
an API that can get candidates. They also provide the classic
downloadable tabular data resources which we do not want to use for
two reasons:

1. We want to use the latest information automatically
2. For EBI it is easier to track use and fund future initiatives

The latter reason is more important than meets the eye. These
resources are *not* free.

* Start with a gene

Let's start with mouse gene [[https://www.wikidata.org/wiki/Q14860079][Shh]] (human alias SHH) and rat [[https://www.wikidata.org/wiki/Q24381323][Brca2]] (human
alias BRCA2). For the human variant the [[https://www.wikidata.org/wiki/Q17853272][wikidata page for BRCA2]] points
to the [[https://www.ncbi.nlm.nih.gov/gene/675][NCBI]] gene description (which is usefully elaborate). That page
in turn points to the [[https://www.uniprot.org/uniprot/P51587][Uniprot]] resource.

Wikidata also has the start and en position of the gene for two
reference genomes. hg38 is known as [[https://www.wikidata.org/wiki/Q20966585][GRCh38.p2]] (2014) which is now at
p13 (2019). So, which one does EBI use? GRCh38.p12 according to the
[[https://www.ebi.ac.uk/gwas/docs/faq][FAQ]]:

:  curl 'https://www.ebi.ac.uk/gwas/rest/api/metadata' -i -H 'Accept: application/json'

#+BEGIN_SRC js
{
  "_embedded" : {
    "mappingMetadatas" : [ {
      "ensemblReleaseNumber" : 97,
      "genomeBuildVersion" : "GRCh38.p12",
      "dbsnpVersion" : 151,
      "usageStartDate" : "2019-07-03T13:00:03.625+0000"
    } ]
  }
}
#+END_SRC

If we look at BRCA2 on p13 the location is
chr13:32315480..32399672 while wikidata has it at p2
chr13:32315086..32973805. In other words we'll need to fetch the
latest somehow.

The [[https://www.ebi.ac.uk/gwas/docs/faq][FAQ]] suggests the API supports searching by gene which should make
a lookup easier. In the full API docs this query is shown

: curl 'https://www.ebi.ac.uk/gwas/rest/api/singleNucleotidePolymorphisms/search/findByGene?geneName=BRCA2' -i -H 'Accept: application/json' > BRCA2.json

results in a list of SNPs with top/curated associations. For summary statistics
there is no lookup by gene yet - I have put in a request. So, for now
we need to figure out the gene position first using the ENSEMBL API.

Wikidata includes an ENSEMBL identifier which makes this relatively
straightforward. In fact, that can even be skipped because ENSEMBL takes
gene [[https://rest.ensembl.org/documentation/info/symbol_lookup][names]]. E.g.

: curl 'https://rest.ensembl.org/lookup/symbol/homo_sapiens/BRCA2?expand=0&format=full' -H 'Content-type:application/json'

#+BEGIN_SRC js
  {"db_type":"core",
   "end":32400266,
   "biotype":"protein_coding",
   "assembly_name":"GRCh38",
   "start":32315086,"object_type":
   "Gene","strand":1,
   "description":"BRCA2 DNA repair associated [Source:HGNC Symbol;Acc:HGNC:1101]",
   "seq_region_name":"13",
   "source":"ensembl_havana",
   "version":15,
   "display_name":"BRCA2",
   "logic_name":"ensembl_havana_gene_homo_sapiens",
   "id":"ENSG00000139618",
   "species":"homo_sapiens"}
#+END_SRC

Note that this uses another reference genome (again!).

Coming to think of it, for summary statistics it should not matter too
much because most associated SNPs will be outside the coding region
anyway. The simple solution is to take anything that is 100Kb around
the gene under study. In gene dense regions this may lead to problems,
but hey, this is a discovery tool. We will use above coordinates to
signify whether a SNP is (likely) outside the gene region.
